packages/@aws-cdk/aws-iotevents/lib/state.ts

```diff
@@ -141,22 +157,32 @@ export class State {
   }
 
   private toStateJson(scope: Construct, actionBindOptions: ActionBindOptions): CfnDetectorModel.StateProperty {
-    const { onEnter } = this.props;
+    const { onEnter, onInput, onExit } = this.props;
     return {
       stateName: this.stateName,
-      onEnter: onEnter && { events: toEventsJson(scope, actionBindOptions, onEnter) },
+      onEnter: {
+        events: toEventsJson(scope, actionBindOptions, onEnter),
+      },
       onInput: {
+        events: toEventsJson(scope, actionBindOptions, onInput),
         transitionEvents: toTransitionEventsJson(scope, actionBindOptions, this.transitionEvents),
       },
+      onExit: {
+        events: toEventsJson(scope, actionBindOptions, onExit),
+      },
     };
   }
 }
 
 function toEventsJson(
   scope: Construct,
   actionBindOptions: ActionBindOptions,
-  events: Event[],
-): CfnDetectorModel.EventProperty[] {
+  events?: Event[],
+): CfnDetectorModel.EventProperty[] | undefined {
+  if (!events) {
+    return undefined;
+  }
+
```

Comment:
Instead of this condition, you can just change `events.map(...` below to `events?.map(...`, and it will have the same effect.